### PR TITLE
Improve call stack on ObtainSnapshot_cancel

### PR DESCRIPTION
**Pull request by markfields**
Clone of #16183, which was closed automatically for being stale. I'm making some tweaks in response to comments over there myself and will work with @Abe27342 to merge.

## Description

Tweaks error surfacing in `getVersions` by regenerating the error stack. This event is a frequent cause of document access issues for partners on odsp--e.g. auth failure gets surfaced via this error. Having a more clear stack that's understandable for application authors seems like a nice value add.

### CODE REVIEW

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts [hunk 1]
```diff
@@ -312,13 +314,29 @@ export class OdspDocumentStorageService extends OdspDocumentStorageServiceBase {
...
+									// they get truncated at this stack frame due to the promise race and how v8 tracks async stack traces--
+									// see https://v8.dev/docs/stack-trace-api#async-stack-traces and the "zero-cost async stack traces" document
+									// linked there. https://v8.dev/blog/fast-async#await-under-the-hood may also be helpful for context on internals.
+									// Regenerating the stack at this level provides more information for logged errors.
+									// Once FF uses an ES2021 target, we could convert the above promise race to use `Promise.any` + AggregateError and
+									// get similar quality stacks with less hand-crafted code.
+									if (err instanceof LoggingError) {
+										const originalStack = err.stack;
+										err.addTelemetryProperties({ originalStack });
+										err.stack = generateStack();
```
[the remark this answers is not in the document]
Comment (markfields) [reply packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:337]: Yeah I was pondering about that last night.  I'm going to leave it pretty manual/explicit for now.  If this works out and more places need this too, we can beef up the helpers in telemetry-utils to streamline it for callers.

[the remark this answers is not in the document]
Comment (Abe27342) [reply packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:337]: cool. If it does end up throwing here, at least we were planning on re-throwing anyway. so seems fine for now

**Review by Abe27342 — APPROVED** [review]

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (markfields) [reply packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:337]: I'm going to use that overwriteStack helper

File: packages/utils/telemetry-utils/src/errorLogging.ts [hunk 2]
```diff
@@ -284,8 +284,13 @@ export function wrapErrorAndLog<T extends LoggingError>(
...
 
-function overwriteStack(error: IFluidErrorBase | LoggingError, stack: string): void {
-	// supposedly setting stack on an Error can throw.
+/**
+ * Attempts to overwrite the error's stack
+ *
+ * There have been reports of certain JS environments where overwriting stack will throw.
+ * If that happens, this adds the given stack as the telemetry property "stack2"
+ */
+export function overwriteStack(error: IFluidErrorBase | LoggingError, stack: string): void {
```
Comment (alexvy86) [inline packages/utils/telemetry-utils/src/errorLogging.ts:293]: Just adding the `@internal` tag. Will be relevant once we complete what @Josmithr has been working on, to not publish `@internal` APIs to our consumers while keeping them available in-repo.

```suggestion
/**
 * Attempts to overwrite the error's stack
 *
 * There have been reports of certain JS environments where overwriting stack will throw.
 * If that happens, this adds the given stack as the telemetry property "stack2"
 *
 * @internal
 */
export function overwriteStack(error: IFluidErrorBase | LoggingError, stack: string): void {
```

**Pull request merged**